File: src/ClientModel/ClassType.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using AutoRest.Core;
using System;
using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    public class ClassType : IType
    {
        public static readonly ClassType Void = new ClassType("java.lang", "Void");
        public static readonly ClassType Boolean = new ClassType("java.lang", "Boolean", defaultValueExpressionConverter: (string defaultValueExpression) => defaultValueExpression.ToLowerInvariant());
        public static readonly ClassType Byte = new ClassType("java.lang", "Byte");
        public static readonly ClassType Integer = new ClassType("java.lang", "Integer");
        public static readonly ClassType Long = new ClassType("java.lang", "Long", defaultValueExpressionConverter: (string defaultValueExpression) => defaultValueExpression + 'L');
        public static readonly ClassType Double = new ClassType("java.lang", "Double", defaultValueExpressionConverter: (string defaultValueExpression) => double.Parse(defaultValueExpression).ToString());
        public static readonly ClassType String = new ClassType("java.lang", "String", defaultValueExpressionConverter: (string defaultValueExpression) => CodeNamer.Instance.QuoteValue(defaultValueExpression));
        public static readonly ClassType Base64Url = new ClassType("com.microsoft.rest.v2", "Base64Url");
        public static readonly ClassType LocalDate = new ClassType("java.time", "LocalDate", defaultValueExpressionConverter: (string defaultValueExpression) => $"LocalDate.parse(\"{defaultValueExpression}\")");
        public static readonly ClassType DateTime = new ClassType("java.time", "OffsetDateTime", defaultValueExpressionConverter: (string defaultValueExpression) => $"OffsetDateTime.parse(\"{defaultValueExpression}\")");
        public static readonly ClassType Duration = new ClassType("java.time", "Duration", defaultValueExpressionConverter: (string defaultValueExpression) => $"Duration.parse(\"{defaultValueExpression}\")");
        public static readonly ClassType DateTimeRfc1123 = new ClassType("com.microsoft.rest.v2", "DateTimeRfc1123", defaultValueExpressionConverter: (string defaultValueExpression) => $"new DateTimeRfc1123(\"{defaultValueExpression}\")");
        public static readonly ClassType BigDecimal = new ClassType("java.math", "BigDecimal");
        public static readonly ClassType UUID = new ClassType("java.util", "UUID");
        public static readonly ClassType Object = new ClassType("java.lang", "Object");
        public static readonly ClassType ServiceClientCredentials = new ClassType("com.microsoft.rest.v2.credentials", "ServiceClientCredentials");
        public static readonly ClassType AzureTokenCredentials = new ClassType("com.microsoft.azure.v2.credentials", "AzureTokenCredentials");
        public static readonly ClassType CloudException = new ClassType("com.microsoft.azure.v2", "CloudException");
        public static readonly ClassType RestException = new ClassType("com.microsoft.azure.v2", "RestException");
        public static readonly ClassType UnixTime = new ClassType("com.microsoft.rest.v2", "UnixTime");
        public static readonly ClassType UnixTimeDateTime = new ClassType("java.time", "OffsetDateTime");
        public static readonly ClassType UnixTimeLong = new ClassType("java.lang", "Long");
        public static readonly ClassType AzureEnvironment = new ClassType("com.microsoft.azure.v2", "AzureEnvironment");
        public static readonly ClassType HttpPipeline = new ClassType("com.microsoft.rest.v2.http", "HttpPipeline");
        public static readonly ClassType Completable = new ClassType("io.reactivex", "Completable");
        public static readonly ClassType AzureProxy = new ClassType("com.microsoft.azure.v2", "AzureProxy");
        public static readonly ClassType RestProxy = new ClassType("com.microsoft.rest.v2", "RestProxy");
        public static readonly ClassType Validator = new ClassType("com.microsoft.rest.v2", "Validator");
        public static readonly ClassType Function = new ClassType("io.reactivex.functions", "Function");
        public static readonly ClassType ByteBuffer = new ClassType("java.nio", "ByteBuffer");
        public static readonly ClassType Resource = new ClassType("com.microsoft.azure.v2", "Resource");
        public static readonly ClassType ProxyResource = new ClassType("com.microsoft.azure.v2", "ProxyResource");
        public static readonly ClassType SubResource = new ClassType("com.microsoft.azure.v2", "SubResource");
        public static readonly ClassType URL = new ClassType("java.net", "URL");
        public static readonly ClassType NonNull = new ClassType("io.reactivex.annotations", "NonNull");
        public static readonly ClassType OperationDescription = new ClassType("com.microsoft.rest.v2", "OperationDescription");
        public static readonly ClassType VoidResponse = new ClassType("com.microsoft.rest.v2", "VoidResponse");
        public static readonly ClassType StreamResponse = new ClassType("com.microsoft.rest.v2", "StreamResponse");
        public static readonly ClassType Context = new ClassType("com.microsoft.rest.v2", "Context");

        public ClassType(string package, string name, IEnumerable<string> implementationImports = null, IDictionary<string,string> extensions = null, bool isInnerModelType = false, Func<string,string> defaultValueExpressionConverter = null)
        {
            Package = package;
            Name = name;
            ImplementationImports = implementationImports;
            Extensions = extensions;
            IsInnerModelType = isInnerModelType;
            DefaultValueExpressionConverter = defaultValueExpressionConverter;
        }

        public string Package { get; }

        public string Name { get; }

        private IEnumerable<string> ImplementationImports { get; }

        private IDictionary<string,string> Extensions { get; }

        public string GetExtensionValue(string extensionKey)
        {
            return Extensions == null || !Extensions.ContainsKey(extensionKey) ? null : Extensions[extensionKey];
        }

        public bool IsInnerModelType { get; }

        private Func<string,string> DefaultValueExpressionConverter { get; }

        public override string ToString()
        {
            return Name;
        }

        public IType AsNullable()
        {
            return this;
        }

        public bool Contains(IType type)
        {
            return this == type;
        }

        public string FullName => $"{Package}.{Name}";

        public void AddImportsTo(ISet<string> imports, bool includeImplementationImports)
        {
            if (Package != "java.lang")
            {
                imports.Add($"{Package}.{Name}");
            }

            if (includeImplementationImports && ImplementationImports != null)
            {
                foreach (string implementationImport in ImplementationImports)
                {
                    imports.Add(implementationImport);
                }
            }
        }

        public string DefaultValueExpression(string sourceExpression)
        {
            string result = sourceExpression;
            if (result != null)
            {
                if (DefaultValueExpressionConverter != null)
                {
                    result = DefaultValueExpressionConverter(sourceExpression);
                }
                else
                {
                    result = $"new {ToString()}()";
                }
            }
            return result;
        }

        public IType ClientType
        {
            get
            {
                IType clientType = this;
                if (this == ClassType.DateTimeRfc1123)
                {
                    clientType = ClassType.DateTime;
                }
                else if (this == ClassType.Base64Url)
                {
                    clientType = ArrayType.ByteArray;
                }
                return clientType;
            }
        }

        public string ConvertToClientType(string expression)
        {
            if (this == ClassType.DateTimeRfc1123)
            {
                expression = $"{expression}.dateTime()";
            }
            else if (this == ClassType.Base64Url)
            {
                expression = $"{expression}.decodedBytes()";
            }

            return expression;
        }

        public string ConvertFromClientType(string expression)
        {
            if (this == ClassType.DateTimeRfc1123)
            {
                expression = $"new DateTimeRfc1123({expression})";
            }
            else if (this == ClassType.Base64Url)
            {
                expression = $"Base64Url.encode({expression})";
            }

            return expression;
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/model/clientmodel/ClassType.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.azure.autorest.model.clientmodel;


import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 The details of a class type that is used by a client.
*/
public class ClassType implements IType {
    public static final ClassType Void = new ClassType("java.lang", "Void");
    public static final ClassType Boolean = new ClassType("java.lang", "Boolean", null, null, false, java.lang.String::toLowerCase);
    public static final ClassType Byte = new ClassType("java.lang", "Byte");
    public static final ClassType Integer = new ClassType("java.lang", "Integer", null, null, false, (String defaultValueExpression) -> defaultValueExpression);
    public static final ClassType Long = new ClassType("java.lang", "Long", null, null, false, (String defaultValueExpression) -> defaultValueExpression + 'L');
    public static final ClassType Double = new ClassType("java.lang", "Double", null, null, false, (String defaultValueExpression) -> java.lang.String.valueOf(java.lang.Double.parseDouble(defaultValueExpression)));
    public static final ClassType String = new ClassType("java.lang", "String", null, null, false, (String defaultValueExpression) -> "\"" + defaultValueExpression + "\"");
    public static final ClassType Base64Url = new ClassType("com.azure.core.implementation", "Base64Url");
    public static final ClassType LocalDate = new ClassType("java.time", "LocalDate", null, null, false, (String defaultValueExpression) -> java.lang.String.format("LocalDate.parse(\"%1$s\")", defaultValueExpression));
    public static final ClassType DateTime = new ClassType("java.time", "OffsetDateTime", null, null, false, (String defaultValueExpression) -> java.lang.String.format("OffsetDateTime.parse(\"%1$s\")", defaultValueExpression));
    public static final ClassType Duration = new ClassType("java.time", "Duration", null, null, false, (String defaultValueExpression) -> java.lang.String.format("Duration.parse(\"%1$s\")", defaultValueExpression));
    public static final ClassType DateTimeRfc1123 = new ClassType("com.azure.core.implementation", "DateTimeRfc1123", null, null, false, (String defaultValueExpression) -> java.lang.String.format("new DateTimeRfc1123(\"%1$s\")", defaultValueExpression));
    public static final ClassType BigDecimal = new ClassType("java.math", "BigDecimal", null, null, false, (String defaultValueExpression) -> java.lang.String.format("new BigDecimal(\"%1$s\")", defaultValueExpression));
    public static final ClassType UUID = new ClassType("java.util", "UUID");
    public static final ClassType Object = new ClassType("java.lang", "Object");
    public static final ClassType ServiceClientCredentials = new ClassType("com.azure.core.credentials", "ServiceClientCredentials");
    public static final ClassType AzureTokenCredentials = new ClassType("com.microsoft.azure.v3.credentials", "AzureTokenCredentials");
    public static final ClassType CloudException = new ClassType("com.microsoft.azure.v3", "CloudException");
    public static final ClassType RestException = new ClassType("com.microsoft.azure.v3", "RestException");
    public static final ClassType UnixTime = new ClassType("com.azure.core.implementation", "UnixTime");
    public static final ClassType UnixTimeDateTime = new ClassType("java.time", "OffsetDateTime");
    public static final ClassType UnixTimeLong = new ClassType("java.lang", "Long");
    public static final ClassType AzureEnvironment = new ClassType("com.microsoft.azure.v3", "AzureEnvironment");
    public static final ClassType HttpPipeline = new ClassType("com.azure.core.http", "HttpPipeline");
    public static final ClassType AzureProxy = new ClassType("com.microsoft.azure.v3", "AzureProxy");
    public static final ClassType RestProxy = new ClassType("com.azure.core.implementation", "RestProxy");
    public static final ClassType Validator = new ClassType("com.azure.core.implementation", "Validator");
    public static final ClassType Function = new ClassType("io.reactivex.functions", "Function");
    public static final ClassType ByteBuffer = new ClassType("java.nio", "ByteBuffer");
    public static final ClassType Resource = new ClassType("com.microsoft.azure.v3", "Resource");
    public static final ClassType ProxyResource = new ClassType("com.microsoft.azure.v3", "ProxyResource");
    public static final ClassType SubResource = new ClassType("com.microsoft.azure.v3", "SubResource");
    public static final ClassType URL = new ClassType("java.net", "URL");
    public static final ClassType NonNull = new ClassType("reactor.util.annotation", "NonNull");
    public static final ClassType OperationDescription = new ClassType("com.azure.core.implementation", "OperationDescription");
    public static final ClassType VoidResponse = new ClassType("com.azure.core.http.rest", "VoidResponse");
    public static final ClassType StreamResponse = new ClassType("com.azure.core.http.rest", "StreamResponse");
    public static final ClassType Context = new ClassType("com.azure.core.util", "Context");


    public ClassType(String package_Keyword, String name, List<String> implementationImports, java.util.Map<String, String> extensions, boolean isInnerModelType) {
        this(package_Keyword, name, implementationImports, extensions, isInnerModelType, null);
    }

    public ClassType(String package_Keyword, String name, List<String> implementationImports, java.util.Map<String, String> extensions) {
        this(package_Keyword, name, implementationImports, extensions, false, null);
    }

    public ClassType(String package_Keyword, String name, List<String> implementationImports) {
        this(package_Keyword, name, implementationImports, null, false, null);
    }

    public ClassType(String package_Keyword, String name) {
        this(package_Keyword, name, null, null, false, null);
    }

    //C# TO JAVA CONVERTER NOTE: Java does not support optional parameters. Overloaded method(s) are created above:
//ORIGINAL LINE: public ClassType(string package, string name, IEnumerable<string> implementationImports = null, IDictionary<string,string> extensions = null, bool isInnerModelType = false, Func<string,string> defaultValueExpressionConverter = null)
    public ClassType(String package_Keyword, String name, List<String> implementationImports, Map<String, String> extensions, boolean isInnerModelType, java.util.function.Function<String, String> defaultValueExpressionConverter) {
        Package = package_Keyword;
        Name = name;
        ImplementationImports = implementationImports;
        Extensions = extensions;
        IsInnerModelType = isInnerModelType;
        DefaultValueExpressionConverter = defaultValueExpressionConverter::apply;
    }

    private String Package;

    public final String getPackage() {
        return Package;
    }

    private String Name;

    public final String getName() {
        return Name;
    }

    private List<String> ImplementationImports;

    private List<String> getImplementationImports() {
        return ImplementationImports;
    }

    private Map<String, String> Extensions;

    private Map<String, String> getExtensions() {
        return Extensions;
    }

    public final String GetExtensionValue(String extensionKey) {
        return getExtensions() == null || !getExtensions().containsKey(extensionKey) ? null : getExtensions().get(extensionKey);
    }

    private boolean IsInnerModelType;

    public final boolean getIsInnerModelType() {
        return IsInnerModelType;
    }

    private java.util.function.Function<String, String> DefaultValueExpressionConverter;

    private java.util.function.Function<String, String> getDefaultValueExpressionConverter() {
        return DefaultValueExpressionConverter;
    }

    @Override
    public String toString() {
        return getName();
    }

    public final IType AsNullable() {
        return this;
    }

    public final boolean Contains(IType type) {
        return this == type;
    }

    public final String getFullName() {
        return java.lang.String.format("%1$s.%2$s", getPackage(), getName());
    }

    public final void AddImportsTo(Set<String> imports, boolean includeImplementationImports) {
        if (!getPackage().equals("java.lang")) {
            imports.add(java.lang.String.format("%1$s.%2$s", getPackage(), getName()));
        }

        if (includeImplementationImports && getImplementationImports() != null) {
            for (String implementationImport : getImplementationImports()) {
                imports.add(implementationImport);
            }
        }
    }

    public final String DefaultValueExpression(String sourceExpression) {
        String result = sourceExpression;
        if (result != null) {
            if (getDefaultValueExpressionConverter() != null) {
                result = DefaultValueExpressionConverter.apply(sourceExpression);
            } else {
                result = java.lang.String.format("new %1$s()", toString());
            }
        }
        return result;
    }

    public final IType getClientType() {
        IType clientType = this;
        if (this == ClassType.DateTimeRfc1123) {
            clientType = ClassType.DateTime;
        } else if (this == ClassType.Base64Url) {
            clientType = ArrayType.ByteArray;
        }
        return clientType;
    }

    public String ConvertToClientType(String expression) {
        if (this == ClassType.DateTimeRfc1123) {
            expression = java.lang.String.format("%s.dateTime()", expression);
        } else if (this == ClassType.Base64Url) {
            expression = java.lang.String.format("%s.decodedBytes()", expression);
        }

        return expression;
    }

    public String ConvertFromClientType(String expression) {
        if (this == ClassType.DateTimeRfc1123) {
            expression = java.lang.String.format("new DateTimeRfc1123(%s)", expression);
        } else if (this == ClassType.Base64Url) {
            expression = java.lang.String.format("Base64Url.encode(%s)", expression);
        }

        return expression;
    }
}